File: RobotTrajectory.lua
--- LUA wrapper for moveit planning environment
-- dependency to tourch.ros
-- @classmod RobotTrajectory

local ffi = require 'ffi'
local torch = require 'torch'
local ros = require 'ros'
local moveit = require 'moveit.env'
local utils = require 'moveit.utils'
local std = ros.std

local RobotTrajectory = torch.class('moveit.RobotTrajectory', moveit)

local f

function init()
  local RobotTrajectory_method_names = {
    "new",
    "delete",
    "release",
    "empty",
    "getGroupName",
    "setGroupName",
    "getWayPointCount",
    "getWayPoint",
    "setWayPointDurationFromPrevious",
    "addSuffixWayPoint",
    "addPrefixWayPoint",
    "insertWayPoint",
    "append",
    "swap",
    "clear",
    "getAverageSegmentDuration",
    "reverse",
    "unwind",
    "findWayPointIndicesForDurationAfterStart",
    "getRobotTrajectoryMsg",
    "setRobotTrajectoryMsg",
    "getWayPoint",
    "getLastWayPoint",
    "getFirstWayPoint",
    "toTensor"
  }
  f = utils.create_method_table("moveit_RobotTrajectory_", RobotTrajectory_method_names)
end

init()

function RobotTrajectory:__init(kinematic_model, group)
  self.o = f.new(kinematic_model:cdata(), group)
  self.moveit_msgs_RobotStateSpec = ros.get_msgspec('moveit_msgs/RobotState')
  self.moveit_msgs_RobotTrajectory = ros.get_msgspec('moveit_msgs/RobotTrajectory')
end

function RobotTrajectory:cdata()
  return self.o
end

function RobotTrajectory:release()
  f.release(self.o)
end

function RobotTrajectory:empty()
  return f.empty(self.o)
end

function RobotTrajectory:getGroupName()
  return ffi.string(f.getGroupName(self.o))
end

function RobotTrajectory:setGroupName(name)
  return f.setGroupName(self.o,name)
end

function RobotTrajectory:getWayPointCount()
  return f.getWayPointCount(self.o)
end

function RobotTrajectory:getWayPoint(index)
  return f.getWayPoint(self.o, index)
end

function RobotTrajectory:setWayPointDurationFromPrevious(index, value)
  f.setWayPointDurationFromPrevious(self.o,index,value)
end

function RobotTrajectory:addSuffixWayPoint(state, dt)
  state = state or moveit.RobotState.createEmpty()
  f.addSuffixWayPoint(self.o,utils.cdata(state),dt)
end

function RobotTrajectory:addPrefixWayPoint(state, dt)
  state = state or moveit.RobotState.createEmpty()
  f.addPrefixWayPoint(self.o,utils.cdata(state),dt)
end

function RobotTrajectory:insertWayPoint(index,state, dt)
  state = state or moveit.RobotState.createEmpty()
  f.insertWayPoint(self.o,index, utils.cdata(state),dt)
end

function RobotTrajectory:append(state, dt)
  state = state or moveit.RobotState.createEmpty()
  return f.append(self.o, utils.cdata(state),dt)
end

function RobotTrajectory:swap(other)
  return f.swap(self.o, utils.cdata(other))
end

function RobotTrajectory:clear()
  return f.clear(self.o)
end

function RobotTrajectory:getAverageSegmentDuration()
  return f.getAverageSegmentDuration(self.o)
end

function RobotTrajectory:reverse()
  f.reverse(self.o)
end

function RobotTrajectory:unwind()
  f.unwind(self.o)
end

function RobotTrajectory:findWayPointIndicesForDurationAfterStart(duration, before, after, blend)
  f.findWayPointIndicesForDurationAfterStart(self.o,duration, before, after, blend)
  return duration, before, after, blend
end

function RobotTrajectory:getRobotTrajectoryMsg(output)
  local msg_bytes = torch.ByteStorage()
  f.getRobotTrajectoryMsg(self.o, msg_bytes:cdata())
  local msg = output or ros.Message(self.moveit_msgs_RobotTrajectory, true)
  msg:deserialize(msg_bytes)
  return msg
end

function RobotTrajectory:setRobotTrajectoryMsg(reference_state, input)
  if torch.isTypeOf(input, ros.Message) then
    local msg_bytes = input:serialize()
    msg_bytes:shrinkToFit()
    f.setRobotTrajectoryMsg(self.o, reference_state, msg_bytes.storage:cdata())
  end
end

function RobotTrajectory:getWayPoint(index,output)
  output = output or moveit.RobotState.createEmpty()
  f.getWayPoint(self.o,index, output:cdata())
  return output
end

function RobotTrajectory:getLastWayPoint(output)
  output = output or moveit.RobotState.createEmpty()
  f.getLastWayPoint(self.o,output:cdata())
  return output
end

function RobotTrajectory:getFirstWayPoint(output)
  output = output or moveit.RobotState.createEmpty()
  f.getFirstWayPoint(self.o, output:cdata())
  return output
end

function RobotTrajectory:toTensor()
  local count = self:getWayPointCount()
  local waypoints
  local robotState
  if count > 0 then
    local tmp = self:getWayPoint(1):getVariablePositions()
    waypoints = torch.zeros(tmp:size(1),count)
    for i = 0,count-1 do
      robotState = self:getWayPoint(i)
      waypoints[{{},i+1}]:copy(robotState:getVariablePositions())
    end
  end
  return waypoints
end

File: src/robot_trajectory.cpp
#include "torch-moveit.h"
#include <moveit/robot_trajectory/robot_trajectory.h>
#include "utils.h"

typedef robot_model::RobotModelPtr RobotModelPtr;
typedef robot_trajectory::RobotTrajectoryPtr RobotTrajectoryPtr;

MOVIMP(RobotTrajectoryPtr*, RobotTrajectory, new)(RobotModelPtr* robot_model, const char *group)
{
   return new RobotTrajectoryPtr(new robot_trajectory::RobotTrajectory(*robot_model, group));
}

MOVIMP(void, RobotTrajectory, delete)(RobotTrajectoryPtr *ptr)
{
  if (ptr)
    delete ptr;
}

MOVIMP(void, RobotTrajectory, release)(RobotTrajectoryPtr *ptr)
{
  ptr->reset();
}

MOVIMP(const char *, RobotTrajectory, getGroupName)(RobotTrajectoryPtr *ptr)
{
  return (*ptr)->getGroupName().c_str();
}

MOVIMP(bool, RobotTrajectory, empty)(RobotTrajectoryPtr *ptr)
{
  return (*ptr)->empty();
}

MOVIMP(void, RobotTrajectory, setGroupName)(RobotTrajectoryPtr *ptr, const char *group_name)
{
  (*ptr)->setGroupName(group_name);
}

MOVIMP(int, RobotTrajectory,getWayPointCount)(RobotTrajectoryPtr *ptr)
 {
 // std::size_t
  return (*ptr)->getWayPointCount();
 }

MOVIMP(void, RobotTrajectory, getWayPoint)(RobotTrajectoryPtr *ptr, int index, RobotStatePtr *out)
 {
  out->reset(new robot_state::RobotState((*ptr)->getWayPoint(index)));
 }

MOVIMP(void, RobotTrajectory, getLastWayPoint)(RobotTrajectoryPtr *ptr, RobotStatePtr *out)
 {
  out->reset(new robot_state::RobotState((*ptr)->getLastWayPoint()));
 }

MOVIMP(void, RobotTrajectory, getFirstWayPoint)(RobotTrajectoryPtr *ptr, RobotStatePtr *out)
 {
  out->reset(new robot_state::RobotState((*ptr)->getFirstWayPoint()));
 }

MOVIMP( void , RobotTrajectory, getWayPointDurations)(RobotTrajectoryPtr *ptr, THDoubleTensor *output)
 {
  std::deque<double> v= (*ptr)->getWayPointDurations();
  THDoubleTensor_resize1d(output, v.size());
  THDoubleTensor* output_ = THDoubleTensor_newContiguous(output);
  std::copy(v.begin(), v.end(), THDoubleTensor_data(output_));
  THDoubleTensor_freeCopyTo(output_, output);
 }

/** @brief  Returns the duration after start that a waypoint will be reached.
  *  @param  The waypoint index.
  *  @return The duration from start; retuns -1.0 if index is out of range.
  */
MOVIMP(double, RobotTrajectory, getWaypointDurationFromStart)(RobotTrajectoryPtr *ptr, std::size_t index)
{
  return (*ptr)->getWaypointDurationFromStart(index);
}


MOVIMP(double, RobotTrajectory, getWayPointDurationFromPrevious)(RobotTrajectoryPtr *ptr,std::size_t index)
 {
 return (*ptr)->getWayPointDurationFromPrevious(index);
 }

MOVIMP(void, RobotTrajectory, setWayPointDurationFromPrevious)(RobotTrajectoryPtr *ptr, int index, double value)
 {
  (*ptr)->setWayPointDurationFromPrevious(index, value);
 }

/**
 * \brief Add a point to the trajectory
 * \param state - current robot state
 * \param dt - duration from previous
 */
MOVIMP(void, RobotTrajectory, addSuffixWayPoint)(RobotTrajectoryPtr *ptr, RobotStatePtr *state, double dt)
{
 (*ptr)->addSuffixWayPoint(**state, dt);
}

MOVIMP(void, RobotTrajectory, addPrefixWayPoint)(RobotTrajectoryPtr *ptr,RobotStatePtr *state, double dt)
 {
  (*ptr)->addPrefixWayPoint(**state, dt);
 }

MOVIMP(void , RobotTrajectory,insertWayPoint)(RobotTrajectoryPtr *ptr,int index, RobotStatePtr *state, double dt)
 {
  (*ptr)->insertWayPoint(index, **state, dt);
 }

 /**
  * \brief Add a trajectory to the end of the current trajectory
  * \param source - the trajectory to append to the end of current trajectory
  * \param dt - time step between last traj point in current traj, and first traj point of new traj
  */
MOVIMP(void, RobotTrajectory, append)(RobotTrajectoryPtr *ptr, const RobotTrajectoryPtr *source, double dt)
{
 (*ptr)->append(**source,dt);
}

MOVIMP(void, RobotTrajectory, swap)(RobotTrajectoryPtr *ptr,const RobotTrajectoryPtr *other)
{
 (*ptr)->swap(**other);
}

MOVIMP(void, RobotTrajectory, clear)(RobotTrajectoryPtr *ptr)
{
 (*ptr)->clear();
}

MOVIMP(double, RobotTrajectory, getAverageSegmentDuration)(RobotTrajectoryPtr *ptr)
{
 return (*ptr)->getAverageSegmentDuration();
}

MOVIMP(void, RobotTrajectory,getRobotTrajectoryMsg)(RobotTrajectoryPtr *ptr, THByteStorage *output)
{
  moveit_msgs::RobotTrajectory trajectory;
 (*ptr)->getRobotTrajectoryMsg(trajectory);

 uint32_t length = ros::serialization::serializationLength(trajectory);
 THByteStorage_resize(output, length + sizeof(uint32_t));
 ros::serialization::OStream stream(THByteStorage_data(output), length + sizeof(uint32_t));
 stream.next((uint32_t)length);
 ros::serialization::serialize(stream, trajectory);
}

MOVIMP(void, RobotTrajectory,setRobotTrajectoryMsg)(RobotTrajectoryPtr *ptr,const RobotStatePtr *reference_state, THByteStorage *serialized_msg)
{
  moveit_msgs::RobotTrajectory rt_msg;
  // deserialize to moveit_msgs::RobotTrajectory message
  long storage_size = THByteStorage_size(serialized_msg);

  uint8_t *data = THByteStorage_data(serialized_msg);

  ros::serialization::IStream stream(data+ sizeof(uint32_t), static_cast<uint32_t>(storage_size-sizeof(uint32_t)));
  ros::serialization::Serializer<moveit_msgs::RobotTrajectory>::read(stream, rt_msg);
  (*ptr)->setRobotTrajectoryMsg(**reference_state, rt_msg);
}


MOVIMP(void, RobotTrajectory, reverse)(RobotTrajectoryPtr *ptr)
{
 (*ptr)->reverse();
}

MOVIMP(void, RobotTrajectory, unwind)(RobotTrajectoryPtr *ptr)
{
 (*ptr)->unwind();
}

/*MOVIMP(void, RobotTrajectory, unwind)(RobotTrajectoryPtr *ptr, const robot_state::RobotState &state)
{
 (*ptr)->unwind(state);
}*/

 /** @brief Finds the waypoint indicies before and after a duration from start.
  *  @param The duration from start.
  *  @param The waypoint index before the supplied duration.
  *  @param The waypoint index after (or equal to) the supplied duration.
  *  @param The progress (0 to 1) between the two waypoints, based on time (not based on joint distances).
  */
MOVIMP(void, RobotTrajectory, findWayPointIndicesForDurationAfterStart)(RobotTrajectoryPtr *ptr,const double &duration, int &before, int &after, double &blend)
{
 (*ptr)->findWayPointIndicesForDurationAfterStart(duration, before, after, blend);
}
